src/plugin.ts

```diff
@@ -194,68 +83,96 @@ export default class DocgenPlugin {
+    this.options = options;
+    this.compilerOptions = compilerOptions;
+    this.docgenOptions = docgenOptions;
 
-          if (!isIncluded(module.userRequest)) {
-            debugExclude(
-              `Module not matched in "include": ${module.userRequest}`
-            );
-            return;
-          }
+    // Maintain the doc gen parser here to save time later
+    this.parser = docGen.withCompilerOptions(compilerOptions, docgenOptions);
+  }
 
-          debugInclude(module.userRequest);
-          modulesToProcess.push(module);
-        });
+  apply(compiler: webpack.Compiler): void {
+    const pluginName = "DocGenPlugin";
 
-        const tsProgram = ts.createProgram(
-          modulesToProcess.map((v) => v.userRequest),
-          compilerOptions
-        );
+    const { exclude = [], include = ["**/**.tsx"] } = this.options;
+    const isExcluded = matchGlob(exclude);
+    const isIncluded = matchGlob(include);
 
-        modulesToProcess.forEach((m) =>
-          processModule(parser, m, tsProgram, {
-            docgenCollectionName,
-            setDisplayName,
-            typePropName,
+    compiler.hooks.compilation.tap(
+      pluginName,
+      (compilation, { normalModuleFactory }) => {
+        compilation.dependencyTemplates.set(
+          // eslint-disable-next-line
+          // @ts-ignore TODO: Figure out why this isn't allowed
+          DocGenDependency,
+          new DocGenDependency.Template({
+            parser: this.parser,
+            docgenOptions: this.docgenOptions,
           })
         );
 
-        cache.save();
-      });
-    });
+        // eslint-disable-next-line
+        // @ts-ignore: TODO: What's the type of a parser?
+        const handler = (parser) => {
+          parser.hooks.program.tap(pluginName, () => {
+            // eslint-disable-next-line
+            // @ts-ignore
+            const { module } = parser.state;
+
+            // Break the recursion. Without this the plugin will get stuck in
+            // an eternal loop as dependencies are added.
+            if (!module.userRequest) {
+              debugExclude(`Ignoring external module: ${module.userRequest}`);
+              return;
+            }
+
+            if (module.external) {
+              debugExclude(`Ignoring external module: ${module.userRequest}`);
+              return;
+            }
+
+            if (!module.rawRequest) {
+              debugExclude(
+                `Ignoring module without "rawRequest": ${module.userRequest}`
+              );
+              return;
+            }
+
+            if (isExcluded(module.userRequest)) {
+              debugExclude(
+                `Module not matched in "exclude": ${module.userRequest}`
+              );
+              return;
+            }
+
+            if (!isIncluded(module.userRequest)) {
+              debugExclude(
+                `Module not matched in "include": ${module.userRequest}`
+              );
+              return;
+            }
```

Comment:
`rawRequest` and `userRequest` should be avoided. Try to use `nameForCondition()` instead.